### PR TITLE
[SymfonyBrdige] Allow to disable command unnesting with configuration

### LINKED ISSUE #51: Allow to disable default FinishesHandlingMessageBeforeHandlingNext
in https://github.com/SimpleBus/SimpleBus/blob/c245f7afa98dcfdb9907e982373323974160744c/Bundle/SymfonyBridge/src/Resources/config/command_bus.yml#L52

FinishesHandlingMessageBeforeHandlingNext is enabled by default, which forces us to use "serialized" execution of the commands. But in our case we put commands to the bus inside the handler and expect that all handlers fail in chain if the inner handler fails, as well as we expect the command to be executed synchronously without delay, as the next code expect the modified state if command executed successfully.

Currently we have to disable this middleware using DI hacks with removing tags from service in order to achieve it.

### PR DESCRIPTION
**Pull request by scaytrase**
Allow disabling command unnesting with `unnest_commands` bundle option. This will prevent `SimpleBus\Message\Bus\Middleware\FinishesHandlingMessageBeforeHandlingNext` from being tagged as a middleware and thus disabling in-memory command queueing

Fixes #51 

cc @cmodijk

### CODE REVIEW

**Comment by cmodijk** [comment]
I'm not sure on the naming of they `unnest_commands` option maybe we should call it `stratergry_..?` with options like `direct` and `queued`. Naming things is hard here @ruudk what do you think?

**Comment by scaytrase** [comment]
> I'm not sure on the naming of they `unnest_commands` option maybe we should call it `stratergry_..?` with options like `direct` and `queued`. Naming things is hard here @ruudk what do you think?

Strategy would be nice if we have a real choice (i.e possibility to introduce and use custom strategy in userland) there. But there is only boolean options. To nest or not to nest.

**Comment by ruudk** [comment]
Why not keeping it simple as `finishes_command_before_handling_next: true`

Then everybody can understand what it means, and you don't need to explain what (un)nesting means.

**Comment by scaytrase** [comment]
> Why not keeping it simple as `finish_command_before_handling_next: true`

I've thought about this, but this way we'll have several ways of loading middlewares - by function (i.e `logging.enabled: true` and by name `finish_command_before_handling_next: true`. I've tried to keep similar approach with existing config.

For me, until the functionality behind the configuration is the same, there will be no real difference, so the final word on naming or config structure is your's to be consistent with future of the library.

**Comment by ruudk** [comment]
What about this:
```yaml
middlewares:
    logger: true
    finishes_command_before_handling_next: true
    other_middlewares: true
```
?

**Comment by scaytrase** [comment]
Looks good. But this will be BC-break?

**Comment by ruudk** [comment]
We can handle that in a BC compatible way. We deprecate the logger option, and propose the new solution, but if you still have the old one set to true, it will just alias that to middlewares.logger or something.

**Comment by scaytrase** [comment]
Ok, got it. I'll fix the PR according to this ideas. 

What should `other_middlewares: false` do? Disable tagged aggregation compiler pass?

File: docs/Symfony/command-bus-bundle.rst [hunk 1]
```diff
@@ -172,6 +172,25 @@ in ``config.yml``:
...
+Messages will be logged to the ``command_bus`` channel with ``%simple_bus.command_bus.logging.level%`` (defaults to ``debug``) level.
+
+Nested commands execution
+-------------------------
+
+By default while calling ``$commandBus->handle($command)`` nested
+in other ``handle`` invocation your command would be delayed and
+pushed to the in-memory command queue in
+``SimpleBus\Message\Bus\Middleware\FinishesHandlingMessageBeforeHandlingNext``
+middleware. If you wan't to utilize explicit command nesting you
```
Comment (ruudk) [inline docs/Symfony/command-bus-bundle.rst]:
```suggestion
middleware. If you want to utilize explicit command nesting you
```

File: docs/Symfony/command-bus-bundle.rst [hunk 2]
```diff
@@ -172,6 +172,25 @@ in ``config.yml``:
...
+can disable it in ``config.yml``:
+
```
Comment (ruudk) [inline docs/Symfony/command-bus-bundle.rst]: I rewrote your sentence a bit to this:
```rst
By default, calls to ``$commandBus->handle($command)`` will not be executed sequentially. Instead, the ``$command`` will be pushed to a in-memory queue in the 
``SimpleBus\Message\Bus\Middleware\FinishesHandlingMessageBeforeHandlingNext``
middleware. Once the handler that triggered the command is finished, the in-memory queue will be processed.

If you don't like this behaviour you can disable it in ``config.yml``:
```

Comment (scaytrase) [reply docs/Symfony/command-bus-bundle.rst]: Replaced with your text, thank you

**Comment by scaytrase** [comment]
> Nice work! Can you maybe also make a functional test that disables the finishes_command_before_handling_next middleware, and executes a command inside a command handler? Maybe check that the execution order is matching.

Sure!

**Comment by scaytrase** [comment]
I've added the test, please review it

**Comment by cmodijk** [comment]
I don't think we should add the option of `other_middlewares` if we want to remove a middleware then we should add options for every specific middleware that we want to disable. Having a disable all options seems weird and can lead to issues down they line. I would prefer if we add options for every specific middleware.

**Comment by scaytrase** [comment]
So I just remove "other_middlewares" option and it's processing?

**Comment by cmodijk** [comment]
@ruudk What do you think?

**Comment by ruudk** [comment]
@cmodijk I agree. Let's remove `other_middlewares` and only make the middlewares configurable that we want to disable right now.

**Comment by cmodijk** [comment]
I approved this PR @ruudk can you merge, split and release this?

**Comment by ruudk** [comment]
Released https://github.com/SimpleBus/SymfonyBridge/releases/tag/v5.1.5

Thanks @scaytrase

**Comment by cmodijk** [comment]
Thanks @ruudk and @scaytrase

**Pull request merged**